internal/liqonet/tunnel-operator/offloaded_pod_controller.go

```diff
@@ -0,0 +1,163 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tunneloperator
+
+import (
+	"context"
+	"sync"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	liqoconsts "github.com/liqotech/liqo/pkg/consts"
+	liqoipset "github.com/liqotech/liqo/pkg/liqonet/ipset"
+	liqoiptables "github.com/liqotech/liqo/pkg/liqonet/iptables"
+	liqovk "github.com/liqotech/liqo/pkg/virtualKubelet/forge"
+)
+
+// OffloadedPodController reconciles an offloaded Pod object
+type OffloadedPodController struct {
+	client.Client
+	liqoiptables.IPTHandler
+	*liqoipset.IPSetHandler
+
+	// Liqo Gateway network namespace
+	gatewayNetns ns.NetNS
+
+	// Local cache of podInfo objects
+	podsInfo *sync.Map
+
+	// Local cache of endpointsliceInfo objects
+	endpointslicesInfo *sync.Map
+}
+
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=core,resources=pods/status,verbs=get;
+
+// NewOffloadedPodController instantiates and initializes the offloaded pod controller.
+func NewOffloadedPodController(cl client.Client, gatewayNetns ns.NetNS, podsInfo *sync.Map, endpointslicesInfo *sync.Map) (*OffloadedPodController, error) {
+	// Create the IPTables handler
+	iptablesHandler, err := liqoiptables.NewIPTHandler()
+	if err != nil {
+		return nil, err
+	}
+	// Create the IPSet handler
+	ipsetHandler := liqoipset.NewIPSetHandler()
+	// Create and return the controller
+	return &OffloadedPodController{
+		Client:             cl,
+		IPTHandler:         iptablesHandler,
+		IPSetHandler:       &ipsetHandler,
+		gatewayNetns:       gatewayNetns,
+		podsInfo:           podsInfo,
+		endpointslicesInfo: endpointslicesInfo,
+	}, nil
+}
+
+// Reconcile pods offloaded from other clusters to the local one
+func (r *OffloadedPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	nsName := req.NamespacedName
+	klog.Infof("Reconcile Pod %q", nsName)
+
+	pod := corev1.Pod{}
+	if err := r.Get(ctx, nsName, &pod); err != nil {
+		err = client.IgnoreNotFound(err)
+		if err == nil {
+			// Pod not found, podInfo object found: delete podInfo object
+			if value, ok := r.podsInfo.LoadAndDelete(nsName); ok {
+				klog.Infof("Pod %q not found: ensuring updated iptables rules", nsName)
+
+				// Soft delete object
+				podInfo := value.(liqoiptables.PodInfo)
+				podInfo.Deleting = true
+				r.podsInfo.Store(nsName, podInfo)
+
+				if err := r.gatewayNetns.Do(r.ensureIptablesRules); err != nil {
+					klog.Errorf("Error while ensuring iptables rules: %w", err)
```

Comment:
You don't need to log it, just create a new error with fmt.Errorf and return it.